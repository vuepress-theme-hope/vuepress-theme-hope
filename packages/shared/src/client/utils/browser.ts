export const checkIsMobile = (ua: string): boolean =>
  /\b(?:Android|iPhone)/i.test(ua);

export const checkIsChromeWebView = (ua: string): boolean =>
  / wv\).+(chrome)\/([\w.]+)/i.test(ua);

export const checkIsSafariMobile = (ua: string): boolean =>
  /version\/([\w.]+) .*mobile\/\w+ (safari)/i.test(ua);

export const checkIsSafari = (ua: string): boolean =>
  /version\/([\w.]+) .*(mobile ?safari|safari)/i.test(ua);

export const checkIsiPhone = (ua: string): boolean =>
  /\((ip(?:hone|od)[\w ]*);/i.test(ua);

export const checkIsiPad = (ua: string): boolean =>
  [
    /\((ipad);[-\w),; ]+apple/i,
    /applecoremedia\/[\w.]+ \((ipad)/i,
    /\b(ipad)\d\d?,\d\d?[;\]].+ios/i,
  ].some((item) => item.test(ua));

export const checkIsWindows = (ua: string): boolean =>
  [
    /microsoft (windows) (vista|xp)/i,
    /(win(?=3|9|n)|win 9x )([nt\d.]+)/i,
    /(windows) nt 6\.2; (arm)/i,
    /(windows (?:phone(?: os)?|mobile))[/ ]?([\d.\w ]*)/i,
    /(windows)[/ ]?([ntce\d. ]+\w)(?!.+xbox)/i,
  ].some((item) => item.test(ua));

export const checkIsIOS = (ua: string): boolean =>
  [
    /ip[honead]{2,4}\b(?:.*os ([\w]+) like mac|; opera)/i,
    /cfnetwork\/.+darwin/i,
  ].some((item) => item.test(ua));

export const checkIsMacOS = (ua: string): boolean =>
  [/(mac os x) ?([\w. ]*)/i, /(macintosh|mac_powerpc\b)(?!.+haiku)/i].some(
    (item) => item.test(ua)
  );
