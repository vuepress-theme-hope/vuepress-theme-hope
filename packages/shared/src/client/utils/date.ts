export const getDate = (
  input: Date | string | number | undefined | null = ""
): Date | null => {
  if (input) {
    if (typeof input === "number") return new Date(input);

    const date = Date.parse(input.toString());

    if (!Number.isNaN(date)) return new Date(date);
  }

  return null;
};
