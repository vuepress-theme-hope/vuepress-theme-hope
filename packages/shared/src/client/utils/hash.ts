export const generateIndexFromHash = (
  content: string,
  total: number,
): number => {
  let hash = 1;

  for (let i = 0; i < content.length; i++) {
    // Jenkins one-at-a-time
    hash += content.charCodeAt(i);
    hash += hash << 10;
    hash ^= hash >> 6;
  }

  hash += hash << 3;
  hash ^= hash >> 11;

  return hash % total;
};
