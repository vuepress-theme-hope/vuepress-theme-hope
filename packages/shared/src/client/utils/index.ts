export * from "./browser";
export * from "./hash";
export * from "./path";
export * from "./message";
export * from "./random";
export * from "./resolveRouteWithRedirect";
