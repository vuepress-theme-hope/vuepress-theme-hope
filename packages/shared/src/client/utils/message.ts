/**
 * Forked and edited from https://github.com/vxhly/vuepress-plugin-one-click-copy/blob/master/bin/Message.js
 *
 * MIT License
 *
 * Copyright (c) 2019 vxhly <pengchengou@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * Â© 2019 GitHub, Inc.
 */

export class Message {
  private containerElement: HTMLElement;
  private messageElements: Record<number, HTMLDivElement> = {};

  // generate or make sure message container element
  constructor() {
    const containerId = "message-container";
    const containerElement = document.getElementById(containerId);

    if (containerElement) this.containerElement = containerElement;
    else {
      this.containerElement = document.createElement("div");
      this.containerElement.id = containerId;
      document.body.appendChild(this.containerElement);
    }
  }

  pop(html: string, duration = 2000): number {
    const messageElement = document.createElement("div");
    const messageId = Date.now();

    messageElement.className = "message move-in";
    messageElement.innerHTML = html;
    this.containerElement.appendChild(messageElement);
    this.messageElements[messageId] = messageElement;

    if (duration > 0)
      setTimeout(() => {
        this.close(messageId);
      }, duration);

    return messageId;
  }

  close(messageId?: number): void {
    if (messageId) {
      const messageElement = this.messageElements[messageId];

      messageElement.className = messageElement.className.replace(
        "move-in",
        ""
      );
      messageElement.className += "move-out";
      messageElement.addEventListener("animationend", () => {
        messageElement.remove();
        delete this.messageElements[messageId];
      });
    } else
      Object.keys(this.messageElements).forEach((id) => this.close(Number(id)));
  }

  destory(): void {
    document.body.removeChild(this.containerElement);
  }
}
