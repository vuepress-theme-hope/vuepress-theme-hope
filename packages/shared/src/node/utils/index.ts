export * from "./date";
export * from "./encode";
export * from "./packageManager";
