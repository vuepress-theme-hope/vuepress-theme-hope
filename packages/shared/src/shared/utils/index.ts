export * from "./assign";
export * from "./dayjs";
export * from "./date";
export * from "./info";
export * from "./url";
