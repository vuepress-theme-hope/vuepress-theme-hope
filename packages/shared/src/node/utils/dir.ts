import { fs, path } from "@vuepress/utils";

export const getDirContents = (dir: string, base = ""): string[] =>
  fs
    .readdirSync(dir)
    .map((item) => {
      if (fs.statSync(path.join(dir, item)).isDirectory())
        return getDirContents(path.join(dir, item), path.join(base, item));

      return path.join(base, item);
    })
    .flat();
