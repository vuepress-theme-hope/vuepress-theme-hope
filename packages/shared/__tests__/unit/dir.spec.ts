/* eslint-disable @typescript-eslint/naming-convention */
import { path } from "@vuepress/utils";
import { expect, it } from "vitest";

import { getDirContents } from "../../src/node/utils/dir.js";

it("should list dir correctly", () => {
  expect(
    getDirContents(path.resolve(__dirname, "./__fixtures__"))
  ).toMatchSnapshot();
});
