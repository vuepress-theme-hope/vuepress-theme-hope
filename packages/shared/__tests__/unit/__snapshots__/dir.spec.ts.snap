// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should list dir correctly 1`] = `
[
  "database.ts",
  "src/README.md",
  "src/component.md",
  "src/cut.md",
  "src/demo.md",
  "src/example.md",
  "src/link.md",
  "theme/empty.ts",
]
`;
